### ISSUE TITLE
.on('notification') handler is not called on IOS

### ISSUE THREAD

**Issue by srameshr** [phonegap/phonegap-plugin-push#1577]
### Expected Behaviour
myPushPlugin.on('notification', function(data) {
  console.log(data.additionalData.extra); // This is called on android but not on IOS
})

### Actual Behaviour

The above should be called on both android and ios

### Reproduce Scenario (including but not limited to)
Just send a push with following JSON format:
```
{
  "data": {
    "title": "",
    "message": "",
    "image": "",
    "extra": {
      "url": ""
    }
  },
  "to": "device-id",
  "notification": {
    title: "",
    body: ""
  },
  "content-available": 1
};
```

This JSON, trigger ```.on('notification')``` on  ```Android``` but not on ```IOS```

#### Steps to Reproduce 

Push registration on client
```
window.pushPlugin = PushNotification.init({
          android: {
            senderID: "myId",
            forceShow: true,
            "icon": "icon",
            "iconColor": "#663467",
            "clearNotifications": false

          },
          ios: {
              senderID: "myId",
              gcmSandbox: false,
              alert: "true",
              badge: "true",
              sound: "true"
          },
        });
```
#### Platform and Version (eg. Android 5.0 or iOS 9.2.1)

  android 6.1.2
  ios 4.3.1

#### (Android) What device vendor (e.g. Samsung, HTC, Sony...)

 Works perfectly on android

#### Cordova CLI version and cordova platform version

    cordova --version                                      6.5.0

#### Plugin version

  phonegap-plugin-push 1.9.2 "PushPlugin"

#### Sample Push Data Payload

```
{
  "data": {
    "title": "",
    "message": "",
    "image": "",
    "extra": {
      "url": ""
    }
  },
  "to": "device-id",
  "notification": {
    title: "",
    body: ""
  },
  "content-available": 1
};
```

**Comment by macdonst** [comment]
@srameshr for GCM on iOS send the regular payload you have above except you don't need title/body twice:

```
{
  "data": {
    "image": "",
    "extra": {
      "url": ""
    }
  },
  "to": "device-id",
  "notification": {
    title: "",
    body: ""
  }
}
```

For GCM on Android is has to be modified somewhat to put everything in `data`.

```
{
  "data": {
    "title": "",
    "body": ""
    "image": "",
    "extra": {
      "url": ""
    }
  },
  "to": "device-id",
}
```

**Comment by srameshr** [comment]
@macdonst Right now I dont have a way to identify the type of device. I will cross check and let you know. Earlier it used to work fine.

**Comment by srameshr** [comment]
@macdonst Is it wrong to send ```notification``` to android or ```title``` and ```body``` under ```data``` to IOS?

**Comment by Jarvey** [comment]
I have kind of the same problem but in my case push.on('notification', function(data) {} is not triggered on android and iOS when the app is in foreground and a notification is received.

ios: 10
android: 5.0.1
phonegap-plugin-push 1.9.2 "PushPlugin"

json iOS:
{
  "app": {
    "alert": {
       "title": "",
       "body": ""
    },
    "badge": "",
    "sound": "",
    "extra": {
      "type": ""
    },
    "notId": ""
};

Android
{
  "data": {
    "title": "",
    "message": "",
    "badge": "",
    "extra": {
      "type": ""
    }
    "notId": ""
};

**Comment by srameshr** [comment]
@Jarvey I followed @macdonst advice. Still it fails to to call .on('notification') on IOS

**Comment by srameshr** [comment]
@Jarvey @macdonst 
```
{
  "data": {
    "title": "",
    "body": ""
    "image": "",
    "extra": {
      "url": ""
    }
  },
  "to": "device-id",
}
```

```.on('notification')``` is failing on android too.

**Comment by srameshr** [comment]
@macdonst Something with the plugin? Because I used to send the same JSON earlier and i just worked out of the box.

**Comment by srameshr** [comment]
@macdonst I updated the plugin to the latest version now. It still fails. ```.on('notification')``` is not being called. Any help here?

**Comment by bardu** [comment]
I experience the same issue. .on('notification') doesn't trigger on both iOS and Android. It has worked at some point in time before.

**Comment by gjongenelen** [comment]
For me the .on(notification) handler is called on android, most of the times...

On iOS, the handler isn't called as soon as the screen turns off.

**Comment by bardu** [comment]
I have commented on the details of my Android issue here:
https://github.com/phonegap/phonegap-plugin-push/issues/820

**Comment by macdonst** [comment]
@srameshr is the app in the foreground or background? Are you clicking on the notification, etc?

**Comment by srameshr** [comment]
@macdonst This has been discussed so many times. Just not for me but for a lot of people, the ```.on('notification')``` handler is not called on IOS. I have tried a lot of workarounds, including
1. Changing the JSON format for GCM on IOS
2. Removing the ```pushRegistration``` on resume to avoid race condition
3. Uploading new certificates to both GCM and IOS
4. And few other stuff

But nothing fixes it. The logs dont indicate any error. And please see <b>I AM TESTING ON A REAL DEVICE AND NOT A SIMULATOR.</b> This keeps popping up again and again on every thread in this repo.

I would suggest you to include an IOS specific ```example``` for both the client and server with GCM as notification provider.

**Comment by Lulun10** [comment]
I have the same issue.

This is my code : 
`$body['aps'] = array(
		'alert' => array(
				'title' => $titre,
				'body' => 'test',
		),
		'sound' => 'default'
);`

Can you help me ?

**Comment by bardu** [comment]
Because of that and I'm using ReactJS anyway I'm investigating to migrate to react-native, there is an FCM module available at
https://github.com/evollu/react-native-fcm

**Comment by srameshr** [comment]
@bardu Even I am doing the same. Migrating my cordova app to React native.

**Comment by srameshr** [comment]
@bardu Yes, even I am doing the same thing. Migrating my cordova app to react native.

**Comment by macdonst** [comment]
@srameshr in what circumstances was `on('notification')` not being called? The payload you were sending should only fire the event when you click on the notification or the app is in the foreground. If you want the app to fire even when the app is in the background you'll want to set the `content-available` property.

```
{
  "to": "device-id",
  "data": {
    "image": "",
    "extra": {
      "url": ""
    }
  },
  "notification": {
    title: "",
    body: ""
  },
  "content_available": true
}
```

**Comment by gs-sramesh** [comment]
@macdonst My app is in the background. It receives a notification. The user clicks on the notification. The app opens. Now I want the ```.on('notification')``` handler to be called. As simple as that. The same flow works perfectly in android. On IOS after the user clicks on the notification and the app opens. But nothing happens after that. I have a simple ```alert``` inside ```.on(notification)```

**Comment by macdonst** [comment]
@gs-sramesh and you see nothing in the logs in XCode?

**Comment by gs-sramesh** [comment]
@macdonst No.

**Comment by macdonst** [comment]
@gr-sramesh I don't know what to tell you. I can't reproduce it. I wonder if it is a race condition where your app doesn't have the notification handler setup quick enough.

**Comment by gs-sramesh** [comment]
@macdonst Il verify few more trials and let you know.

**Comment by bardu** [comment]
@macdonst Maybe we need a clarification in the payload doc (https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md) about the form of the `content available` property for both iOS and Android.

In the doc there are actually two forms of the property `content_available` and `content-available`, so which is the right one?

For the value of this property there are `"1"`,  `1` and `true`, which is the proper one?

**Comment by lfreneda** [comment]
Any fix? I'm facing the same issue.

**Comment by KeaganFouche** [comment]
Hey, I have a solution...

There's a bit of confusion here about push notifications and Ionic, the documentary can be a little confusing.

When you import "Push" there are two kinds, 
1) `import { Push, PushObject, PushOptions } from '@ionic-native/push';` (From [Ionic](https://ionicframework.com/docs/native/push/) )
2) `import { Push, PushToken } from '@ionic/cloud-angular';` (also from [Ionic](https://docs.ionic.io/services/push/#injecting-push) )

Option 1 is the native code, where you will need to use "this.push.on('notification' ...)".
Option 2 is the angular wrapper, which ties everything up for you. You won't need to call "this.push.finish()" either, the [wrapper will take care of it.](https://docs.ionic.io/services/push/#handling-notifications)
The code which worked for me (ios and android):

`if (!platform.is('cordova')) { //if you are using a computer or simulator
                let alert = this.alertController.create({
                    title: "No FCM token",
                    subTitle: "A fake token has been assigned.",
                    buttons: ['OK']
                });
                alert.present();

                console.warn("Push notifications not initialized. Cordova is not available - Run in physical device");
                localStorage.setItem('pushNotificationToken', "THIS_IS_A_FAKE_TOKEN");
            }
            else if (platform.is('cordova')) {
                console.log("Calling register");
                this.register(); 
            }
        });
    }

    register() {
        this.push.register() //call to FCM/APNS to get your register token
            .then((t: PushToken) => {
                let alert = this.alertController.create({
                    title: "Fcm Token Assigned",
                    subTitle: t.token,
                    buttons: ['OK']
                });
                alert.present();              
                localStorage.setItem('pushNotificationToken', t.token);
                return this.push.saveToken(t); //save token. NB this is a method of Push from ionic/cloud-angular
            })
            .then((t: PushToken) => {
                console.log('Token saved');
            });
        
        this.push.rx.notification() //subscribe for notifications. This is the wrapper which handles everything for you, when a notification comes through this will run.
            .subscribe((msg) => {
                let alert = this.alertController.create({
                    title: "App in foreground?",
                    subTitle: "Was the app in the foreground when this notification came through? " + msg.raw.additionalData.foreground, //you could use this kinda functionality to adjust how to handle notifications appropriately.
                    buttons: ['OK']
                });
                alert.present();

                let alert = this.alertController.create({
            title: msg.title,
            subTitle: msg.text,
            buttons: ['OK']
        });
        alert.present();
            });
    }`


Also note, for IOS you will need a p12 certificate for push notifications. Get one through following these [links.](https://docs.ionic.io/services/profiles/#ios-push-certificate)
After that, you'll need to enable Push Notifications in App Services and optionally remote notifications to allow background notifications for IOS.
Personally IOS is a pain, but once you've learned all the shenanigans you'll be fine. Please read the full documents that I have linked so that you can understand the processes.

The boring stuff (specs)
My version specs:
Cordova 7.0.1
Node: 7.9.0
Ionic 2.2.3 (At present, Ionic V3.5.0 is out but I've found it buggy)
Typescript: 2.2.1
Ionic/app-scripts: 1.3.0

Good luck!

Let me know if you need anything or if I'm wrong about anything.
Happy coding!

**Comment by andymurakami** [comment]
@gs-sramesh  @srameshr do you found any solution ? I have the same problem.

**Comment by pawan-mittal** [comment]
@macdonst I have the same problem on iOS  push.on('notification', function(data) {} is not triggered when the app is in foreground and a notification is received. Also, as I receive notification and clicking on it as well doesn't fire push.on('notification', function(data) {}.

**Comment by codeunity** [comment]
No fix? I have the same issue. on('notifcation') Doesnt get called. In the logs i can see Notification received.

**Comment by adidamty** [comment]
Hi,

Me too ... Same issue.
this is frustrating.

**Comment by pawan-mittal** [comment]
For me - I was using both Local Notification plugin (https://github.com/katzer/cordova-plugin-local-notifications) and Push Plugin in my app which caused this issue. Issue was caused by Local Notification plugin. So after removing it and using another local notification plugin (git+https://github.com/Telerik-Verified-Plugins/LocalNotification.git), my issue got resolved. No issue on Push Plugin side.

**Comment by andymurakami** [comment]
I´m not use Local Notification, only Push Plugin, and does not work.  =(

**Comment by adidamty** [comment]
@pawan-mittal @andymurakami thank you for your facst response.
I am not using Local Notification plugin onlt phonegap-plugin-push version 2.1.2. 
The plugin working with Android, and not working with real device IOS 11.2.5 ( iphone 6 ) - on notification is not fired, but i am getting regestration id using PushSharp, and also PushSharp event is fired - ApnServiceBroker_OnNotificationSucceeded.
Are you able to get notification with phonegap-plugin-push ? which version are you using ? did you did anything special to xcode project beside build ionic cordova with profice dev ?

**Comment by andymurakami** [comment]
I have the same issue. the 'on notification' event is not fired when push notification is taped. Sorry =(

**Comment by pawan-mittal** [comment]
I am using phonegap-plugin-push version ^2.0.0. I have tested using iPhone 5S, iPhone 7. In Xcode, enabled Push Notification, also enabled 'Remote notifications' in Background Modes in Capabilities.

**Comment by pawan-mittal** [comment]
If are having Registration token then you can test Push Notification using http://apns-gcm.bryantan.info/

**Comment by adidamty** [comment]
Hi ,

thank you both for your assisatnce.
@pawan-mittal - 
1. I already enabled Push notification. i don't remember i enable "Remote notifications" on capabilities, i am building now project again and check this. did you succeed to get notification with OS 10 and above ?
Also from link  http://apns-gcm.bryantan.info/ i am getting error when trying to send notification 
Error: Unable to connect to 'ssl://gateway.push.apple.com:2195': (0) 
what does it mean ?

**Comment by adidamty** [comment]
Hi

@pawan-mittal - do i need to add googleService-info.plist to my xcode project ? maybe this is the issue ?

**Comment by pawan-mittal** [comment]
@adidamty Yes I was able to make get notification working iOS 10 and above. I think you have some issue with set up or installation of plugin. Plugin is stable, I have been using it for long time.

**Comment by pawan-mittal** [comment]
Push Plugin installation required CocoaPods, so you need to use .xcworkspace instead of .xcodeproj.

**Comment by pawan-mittal** [comment]
you don't need googleService-info.plist for Push Notification for iOS.

**Comment by pawan-mittal** [comment]
Are you using Ionic or standalone Phonegap/Cordova to create App ?

**Comment by adidamty** [comment]
Hi,

@pawan-mittal - I know :
1. i install cocoapods after remove it all, and remove also repository :
    sudo rm -fr ~/.cocoapods/repos/master
2. Than i install cocapods :
   sudo gem install cocoapods --version 1.2.0
3. install cocoapods repos :
   pod setup
4. run pod install ( from platforms\ios )

I am using ionic cordova 👍 
ionic cordova build ios --profile=dev --release  

something missing and it's so fustratring.

thank you,
Adi Damty

**Comment by adidamty** [comment]
@pawan-mittal 
also using .xcworkspace ( with pods ) and not .xcodeproj

**Comment by pawan-mittal** [comment]
Are you using ionic 2? If yes then I have created re-usable provider you can use that. I can share file.
Your steps seems fine.

**Comment by adidamty** [comment]
@pawan-mittal 
yes i am using ionic 2. i will apricate this, i am fighting with it for couple of days.

**Comment by pawan-mittal** [comment]
Steps to use the PushNotification Provider - 
1. Import 'import { PushNotificationProvider } from '../providers/push-notification/push-notification';
' in app.component.ts.
2. Once device is ready - call this.pushNotificationUtil.initializePush();
[push-notification.ts.txt](https://github.com/phonegap/phonegap-plugin-push/files/1665177/push-notification.ts.txt)

Ref for more details - https://ionicframework.com/docs/native/push/

**Comment by pawan-mittal** [comment]
@adidamty let me know if any help required.

**Comment by adidamty** [comment]
Hi,

@pawan-mittal - thank you for your assistance !
in first glance it seems to see the only differnet is with options -  mine   badge: false, and yours    badge:'true' .   
Also i have logger that send info to DB, i am seeing the message for registration but not seeing any log\console messages on notification.

**Comment by adidamty** [comment]
@pawan-mittal i will check and update. thank you.

**Comment by adidamty** [comment]
@pawan-mittal which version of cocapods you are using ?

**Comment by pawan-mittal** [comment]
@adidamty cmd: gem which cocoapods - 
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.3.1/lib/cocoapods.rb

**Comment by adidamty** [comment]
@pawan-mittal
i used cocapods 1.2 now remove cocapods and cocapods repository and installed 1.3.1.
will update soon. 
thank you.

**Comment by adidamty** [comment]
@pawan-mittal 
Same issue - cocapods 1.3.1 and still on notification event is not fired.
I also make sure both Push notification and remote notifications are enabled.

**Comment by adidamty** [comment]
@pawan-mittal 
Hi, 

I found the issue, the pem file i created was invalid. 
here is how pem file need to be created :

1. Create Certificate .pem from apn_development.cer.
   Run terminal with command :
   openssl x509 -inform der -in aps_development.cer -out apns-dev-cert.pem

2. Create Key .pem from Key .p12.
   Run terminal with command :
   openssl pkcs12 -nocerts -in ApplePushNotificationDevelopmentSSL.p12 -out apns-dev-key.pem 
       
3. Remove pass phrase asked in step 12 :
   Run terminal from desktop\keys with command :
   openssl rsa -in apns-dev-key.pem -out apns-dev-key-noenc.pem   

4. Now we have to merge the Key .pem and Certificate .pem to get Development .pem needed for  
   Push Notifications in Development Phase.
   Run terminal with command :
   cat apns-dev-cert.pem apns-dev-key-noenc.pem > apns-dev.pem

5. Go to http://apns-gcm.bryantan.info/ and test certificate ( use file apns-dev.pem )

Thank you for your assistance !

**Comment by pawan-mittal** [comment]
@adidamty thanks yes above steps are for .pem file , I was assuming certs and .Pem will be valid one 🙂

**Comment by andymurakami** [comment]
I´m using Ionic 1, and the notifications sent by Firebase Console, receive ok but not start 'on notification' event on tap.

**Comment by stellmaurer** [comment]
I am having this issue as well. The .on('notification') function was firing for both Android and iOS, but now it only fires on Android. I'm using Ionic 3 with the latest version of Cordova and the latest version of this plugin (2.2.3).

**Comment by stellmaurer** [comment]
So, I removed the plugin "cordova-plugin-local-notification" and it is now working on iOS again, so there is a compatibility issue with these two plugins. I was using the beta version of that plugin (0.9.0-beta.2), which might have been the issue. I will do some more testing and post back with more information.

**Comment by stellmaurer** [comment]
So, it turns out that phonegap-plugin-push and cordova-plugin-local-notification are just not fully compatible with each other.

https://github.com/phonegap/phonegap-plugin-push/issues/2266#issuecomment-415729150
